gtest/src/wasm_executor.rs

```diff
@@ -0,0 +1,349 @@
+            .expect("Failed call");
+
+        let execution_result = match prt_to_result_array[0] {
+            Val::I32(ptr_to_result) => self.read_result(ptr_to_result),
+            _ => {
+                panic!("{}", "Got wrong type")
+            }
+        };
+        (execution_result, self.get_memory_updates())
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn call_void_function(&mut self, function_name: &str) -> Vec<JournalNote> {
+        let function = self.get_function(function_name);
+        function
+            .call(&mut self.store, &[], &mut [])
+            .expect("Failed call");
+
+        self.get_memory_updates()
+    }
+
+    fn get_memory_updates(&mut self) -> Vec<JournalNote> {
+        let mut pages_data = BTreeMap::new();
+        let mut buffer = [0u8; PageNumber::size()];
+        let memory_size = self.memory.data_size(&mut self.store);
+        for page_number in (0..memory_size).step_by(PageNumber::size()) {
+            if let Err(err) = self.memory.read(&mut self.store, page_number, &mut buffer) {
+                panic!("{:?}", err.to_string())
+            }
+            pages_data.insert(PageNumber::new_from_addr(page_number), buffer.to_vec());
+        }
+        self.build_jornal_notes(pages_data)
+    }
+
+    fn build_jornal_notes(&self, pages_data: BTreeMap<PageNumber, Vec<u8>>) -> Vec<JournalNote> {
+        pages_data
+            .into_iter()
+            .map(|(page_number, data)| JournalNote::UpdatePage {
+                program_id: self.program_id,
+                page_number,
+                data: PageBuf::new_from_vec(data).expect("Failed to convert data to page buffer"),
+            })
+            .collect()
+    }
+
+    fn get_function(&mut self, function_name: &str) -> Func {
+        self.instance
+            .get_func(&mut self.store, function_name)
+            .expect("No function with such name")
+    }
+
+    fn read_result(&mut self, ptr_to_result_data: i32) -> Vec<u8> {
+        let mut ptr_to_result_buffer: Vec<u8> = Vec::new();
+        let buffer_size = ptr_to_result_data.to_be_bytes().len();
+        ptr_to_result_buffer.resize(buffer_size, 0);
+
+        self.memory
+            .read(
+                &self.store,
+                ptr_to_result_data as usize,
+                &mut ptr_to_result_buffer,
+            )
+            .expect("Failed to read data ptr");
+
+        let mut result_len_buffer = [0u8];
+
+        self.memory
+            .read(
+                &self.store,
+                ptr_to_result_data as usize + buffer_size,
+                &mut result_len_buffer,
+            )
+            .expect("Failed to read data length");
+
+        let decoded_ptr_to_result =
+            i32::from_ne_bytes(ptr_to_result_buffer.try_into().unwrap()) as usize;
+        let result_len = result_len_buffer[0] as usize;
+
+        let mut results = Vec::new();
+        results.resize(result_len, 0u8);
```

Comment:
As an option (using arrays instead of Vecs allows minimize usage of the heap).

```rust
fn read_result(&mut self, ptr_to_result_data: i32) -> Vec<u8> {
    let offset = ptr_to_result_data as usize;
    let mut ptr = [0_u8; mem::size_of::<i32>()];

    self.memory
        .read(&self.store, offset, &mut ptr)
        .expect("Failed to read data ptr");

    let mut len = [0_u8; mem::size_of::<i32>()];

    self.memory
        .read(&self.store, offset + ptr.len(), &mut len)
        .expect("Failed to read data length");

    let ptr = i32::from_ne_bytes(ptr) as usize;
    let len = i32::from_ne_bytes(len) as usize;

    let mut result = vec![0; len];

    self.memory
        .read(&self.store, ptr, &mut result)
        .expect("Failed to read result");

    result
}
```